### ISSUE TITLE
Type inference and generics when dealing with default `nil`

### ISSUE THREAD

**Issue by luislavena** [crystal-lang/crystal#2456]
Given the following example:

``` crystal
class Foo(T)
  @value : T

  def initialize(@value : T = nil)
  end
end

a = Foo.new
pp typeof(a.@value)
```

It produces the expected output:

```
typeof(a.@value) = Nil
```

However, Attempt to use `T?` for type instead indicates that type cannot be inferred, even when default value already states `nil`:

```
Error in ./4.cr:8: can't infer the type parameter T for the generic class Foo(T). Please provide it explicitly

a = Foo.new
        ^~~
```

Of course it works if any value is provided to `Foo.new`, but was expecting that `T?` was reduced to just `Nil` instead.

Thank you.

**Comment by asterite** [comment]
You mean it should infer that `T` is `Nil` in this case?

**Comment by asterite** [comment]
Mmm... what's the failing code? I don't understand if you place `T?` in the type declaration, in the argument or in both.

**Comment by luislavena** [comment]
@asterite there are a couple of failures:

#### 1. Unable to infer type when default do not define type (but was defined in the class body:

```crystal
class Foo(T)
  @value : T

  def initialize(@value = nil)
  end
end

a = Foo.new
pp typeof(a.@value)
```

Error:

```
Error in ./7.cr:8: can't infer the type parameter T for the generic class Foo(T). Please provide it explicitly

a = Foo.new
        ^~~
```

Error goes away when defined:

```crystal
class Foo(T)
  @value : T

  def initialize(@value : T = nil)
  end
end

a = Foo.new
pp typeof(a.@value)
```
```
typeof(a.@value) = Nil
```

#### 2. Type is not detected when default value is provided for `T?`

```crystal
class Foo(T)
  @value : T?

  def initialize(@value : T? = nil)
  end
end

a = Foo.new
pp typeof(a.@value)
```

Error:

```
Error in ./7.cr:8: can't infer the type parameter T for the generic class Foo(T). Please provide it explicitly

a = Foo.new
        ^~~
```

Solved doing `Foo(Nil).new` or `Foo.new(nil)` but defeats the purpose of having `T?` defined and `nil` as default value.

```
typeof(a.@value) = Nil
```

Perhaps I'm doing something wrong? (aka: using it wrong) :trollface:

**Comment by asterite** [comment]
I think case number 1 is OK to error, as you don't put a type restriction in the method argument (and the compiler won't add it just because the instance variable has a type restriction).

The second case should work... I'll investigate, but it's low priority right now because there's a simple workaround:

```crystal
class Foo(T)
  def self.new
    new(nil)
  end

  def initialize(@value : T?)
  end
end

a = Foo.new
pp typeof(a.@value)
```

**Comment by luislavena** [comment]
No worries about first scenario.

For the second scenario, it was the one baffling me for 45 minutes before reporting this :stuck_out_tongue_winking_eye: 

I'm afraid that overriding `self.new` might not work when having multiple options, so for the time being will simply pass `nil` as second parameter :wink: 

Thank you for looking into this.

Cheers.

**Comment by sdogruyol** [comment]
@luislavena as you may already remember i've also passed `nil` to surpass this issue for `Radix`.

https://github.com/sdogruyol/radix/commit/8f62dadc9210595208713bf944fd446c72241034#diff-931326af4420bfa50aed1c3d3376a877R32

**Comment by luislavena** [comment]
@sdogruyol correct, but as confirmed by Ary, there is a bug on handling the default value for type inference in this case.

Radix suffering this issue is just one of a few internal projects affected by this.

**Comment by asterite** [comment]
I'm still not sure about fixing this. Right now we have this:

```crystal
def foo(x : T?)
  T
end

puts foo(1) # => Int32

a = rand() > 0.5 ? 1 : nil
puts foo(a) # => Int32
```

The reason is that the restriction `T?` is actually `Nil | T`.

In the first case `1` doesn't match `Nil`, but `T` does so `T` becomes Int32.

In the second case `a` is of type `Nil | Int32` (the compiler always puts `Nil` as a first member of a union`), so it matches `Nil | Int32` against `Nil` and it gets `Nil`. `Nil` is then discarded from the restriction and we are left with `T`. Then `Int32` is matched against `T` and it becomes `Int32`.

If we try this:

```crystal
def foo(x : T | Int32)
  T
end

a = rand() > 0.5 ? 1 : "hello"
puts foo(a)
```

we get an error, because the type of `a`, `Int32 | String`, will match against `T` and then `String` won't match against `Int32`.

In conclusion, the current behaviour is bad and it's probably a hack we added to remove `Nil` from a type, I think in `compact` or similar. Free variables should probably never match inside a union type, and in this process types shouldn't be discarded from a union.

Going back to the original example:

```crystal
class Foo(T)
  @value : T?

  def initialize(@value : T? = nil)
  end
end

Foo.new(1)
```

the above shouldn't match, because using `T?` as a type restriction shouldn't make `T` match as a free variable.

The correct way to implement this is:

```crystal
class Foo(T)
  @value : T | Nil

  def initialize(@value : T = nil)
  end
end

p Foo.new(1)
p Foo.new
```

Now both `nil` and `1` can match `T` and the generic type can be deduced, but we say that `@value` is a nilable T.

**Comment by luislavena** [comment]
Thank you @asterite for all the details.

I think the confusion is generated by assuming `T?` and `T | Nil` are the same, yet they aren't.

**Comment by asterite** [comment]
They are the same. The problem is that the compiler will use `T` as a free variable even if it's inside `T | NIl` (or `T?`), and I'm not sure that's a desired behaviour.

**Comment by bcardiff** [comment]
I don't think that the last example of @asterite is right:

```crystal
class Foo(T)
  @value : T | Nil

  def initialize(@value : T = nil)
  end
end
```

On a `Foo(Int32)`, the `T=Int32`, so the default value of `nil` does not make sense with the type restriction. Unless we make the default value "append" a `typeof(default_value)` to the type restrictions automatically.

I would expect the following be the expected implementation

```crystal
class Foo(T)
  @value : T?

  def initialize(@value : T? = nil)
  end
end
```

The thing is how to match `Int32` with `T | Nil`. Instead of trying to make them **equal**, I would say we need to make it **substitutable** `Int32 <: T | Nil`, hence _yes_ if `T = Int32`

**Comment by luislavena** [comment]
Thank you @asterite and @bcardiff for the responses. Maybe is because is Friday and is an indication I should be doing something else by this time, but kinda got lost somewhere :confused: 

Please bear with me the following details.

Given the following code:

```crystal
class Foo(T)
  @value : T | Nil

  def initialize(@value : T = nil)
  end
end

a = Foo(Int32).new
pp typeof(a)              # => Foo(Nil)
pp typeof(a.@value) # => Nil 
```

I would have expected that `Foo(Int32).new` to make `@value` of type `Int32?` and not `Nil` and `a` be `Foo(Int32)`, as defined.

This seems to be honored if `T | Nil` is also used in the initializer:

```crystal
class Foo(T)
  @value : T | Nil

  def initialize(@value : T | Nil = nil)
  end
end

a = Foo(Int32).new
pp typeof(a)        # => Foo(Int32)
pp typeof(a.@value) # => Int32?

b = Foo.new(:root)
pp typeof(b)        # => Foo(Symbol)
pp typeof(b.@value) # => Symbol?
```

This also works if `T?` is used in both places (as indicated, both `T | Nil` and `T?` are the same).

Which allows proper usage of *nilable* `@value` and allow the following things to work:

```crystal
class Foo(T)
  @value : T | Nil

  def initialize(@value : T? = nil)
  end
end

class Bar(T)
  @root : Foo(T)

  def initialize
    @root = Foo(T).new
  end

  def change!(value : T)
    @root = Foo(T).new(value)
  end
end

a = Bar(Int32).new
pp typeof(a.@root)        # => Foo(Int32)
pp typeof(a.@root.@value) # => Int32?
pp a.@root.@value         # => nil

a.change!(100)
pp typeof(a.@root)        # => Foo(Int32)
pp typeof(a.@root.@value) # => Int32?
pp a.@root.@value         # => 100
```

I believe my confusion was generated when the compiler allowed me to define `@value` type `T | Nil`, yet in `initialize`, only `T` was indicated, turning things a bit inconsistent.

So not sure if the following code should either warn or not compile?

```crystal
class Foo(T)
  @value : T | Nil

  def initialize(@value : T = nil)
  end
end
```

Thank you for your patience going over this and to my lack of understanding.

Cheers.

**Comment by asterite** [comment]
@luislavena That the first snippet types it as `Foo(Nil)` is a bug, it should be `Foo(Int32)`.

**Comment by luislavena** [comment]
Hello @asterite, I believe this issue has been solved in recent versions:

```crystal
class Foo(T)
  @value : T | Nil

  def initialize(@value : T = nil)
  end
end

a = Foo(Int32).new
pp typeof(a)
pp typeof(a.@value)
```

Output:

```
typeof(a) # => Foo(Int32)
typeof(a.@value) # => (Int32 | Nil)
```

Originally `Foo(Nil)` was reported and type of `@value` was `Nil`.

Based on that I'm closing this.

Yay! :smiley:

**Comment by asterite** [comment]
The last code doesn't compile anymore...

**Comment by Sija** [comment]
@asterite play.crystal-lang.org shows it stopped working in Crystal v0.20.5.

**Comment by asterite** [comment]
It's not something that I will fix any time soon, given that the workaround is so simple (create an empty `new` method that invokes the other `new` with the default argument).

**Comment by Sija** [comment]
Below code (notice `forall T` in `initialize`) works well in all recent versions though:

```cr
class Foo(T)
  @value : T | Nil

  def initialize(@value : T = nil) forall T
  end
end

a = Foo(Int32).new
pp typeof(a)
pp typeof(a.@value)
```

**Comment by asterite** [comment]
@Sija I think not, `Foo.new` fails

**Comment by Sija** [comment]
@asterite That's strange, I'm getting output like below:

```cr
typeof(a) # => Foo(Int32)
typeof(a.@value) # => (Int32 | Nil)
```

https://play.crystal-lang.org/#/r/2pux

**Comment by asterite** [comment]
I said `Foo.new`, not `Foo(Int32).new` :-)

**Comment by Sija** [comment]
@asterite oops, my bad :)
